### PR TITLE
Bugfix/immutable states

### LINKED ISSUE #57: Ensure State immutability and fix reference bugs in Agents

### PR DESCRIPTION
**Pull request by wujuu**
Closes #57

### CODE REVIEW

**Comment by wujuu** [comment]
@wujuu  add tuples as immutable lists

File: environments/tic_tac_toe/tic_tac_toe_engine_utils.py [hunk 1]
```diff
@@ -170,4 +144,4 @@ def __init__(self, message):
         self.message = message
 
     def __str__(self):
-        return str(self.message)
+        return str(self.message)
```
Comment (JanKapala) [inline environments/tic_tac_toe/tic_tac_toe_engine_utils.py]: non-empty endline

File: reinforcement_learning/agents/q_learning_agent/q_learning_agent.py [hunk 2]
```diff
@@ -55,10 +50,9 @@ def restart(self):
         self._reset_episode_info()
 
     def _update(self, new_state):
-        new_state = copy.deepcopy(new_state)
-        prev_state = copy.deepcopy(self._prev_state)
-        prev_action = copy.deepcopy(self._prev_action)
-        prev_reward = copy.deepcopy(self._prev_reward)
+        prev_state = self._prev_state
+        prev_action = self._prev_action
+        prev_reward = self._prev_reward
```
Comment (JanKapala) [inline reinforcement_learning/agents/q_learning_agent/q_learning_agent.py]: chain variables

File: environments/tic_tac_toe/tic_tac_toe_engine_utils.py (hunk 1 above)
Comment (JanKapala) [reply environments/tic_tac_toe/tic_tac_toe_engine_utils.py]: non-empty endline

File: reinforcement_learning/agents/q_learning_agent/q_learning_agent.py (hunk 2 above)
Comment (JanKapala) [inline reinforcement_learning/agents/q_learning_agent/q_learning_agent.py]: chain variables

**Pull request merged**